Adopt SIMD in attribute and text scanning

Constellation opened the pull request:
#### e8401be1912bd7ccd878b2605d7edf2c91cf74f4
<pre>
Adopt SIMD in attribute and text scanning
<a href="https://bugs.webkit.org/show_bug.cgi?id=273977">https://bugs.webkit.org/show_bug.cgi?id=273977</a>
<a href="https://rdar.apple.com/127843610">rdar://127843610</a>

Reviewed by Mark Lam.

This patch integrates SIMD into attribute value and string scanning.
We also attach UNLIKELY / LIKELY to bailout paths.

* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::parseCompleteInput):
(WebCore::HTMLFastPathParser::scanText):
(WebCore::HTMLFastPathParser::scanTagName):
(WebCore::HTMLFastPathParser::scanAttributeValue):
(WebCore::HTMLFastPathParser::parseChildren):
(WebCore::HTMLFastPathParser::parseAttributes):
(WebCore::HTMLFastPathParser::parseContainerElement):

Canonical link: <a href="https://commits.webkit.org/278647@main">https://commits.webkit.org/278647@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/412f72f5c3d62e0cae9ab6c160dd1327e0d5ec26

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51077 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3408 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54334 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1766 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53379 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36669 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1441 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41592 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53176 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28010 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44017 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22711 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25337 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1271 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9517 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/44416 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47317 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1345 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55929 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50579 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26187 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1239 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48992 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27435 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44085 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48122 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28315 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/58053 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7447 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27166 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11942 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

ddegazio reviewed (approved):
LGTM!

hyjorc1 commented on Source/WTF/wtf/text/StringParsingBuffer.h:
```diff
@@ -53,7 +53,7 @@ class StringParsingBuffer final {
     constexpr void setPosition(const CharacterType* position)
     {
-        ASSERT(!m_data.empty());
+        ASSERT(position <= m_data.data() + m_data.size());
```
If `m_data` is empty, then `m_data.data() + m_data.size()` is `nullptr + 0`. Maybe it's fine. But performing arithmetic on a nullptr can lead to undefined behavior or incorrect assertions.

I think we should do
```
ASSERT(!m_data.empty());
ASSERT(position >= m_data.data() && position <= m_data.data() + m_data.size());
```

Constellation replied:
@hyjorc1 This is not correct. The original code was the above, and it gets broken when `std::span` gets integrated and changed to `ASSERT(!m_data.empty());`, this is completely wrong since we should accept position change even if `m_data` is empty.

Constellation replied:
https://github.com/WebKit/WebKit/commit/d2a029f757b316ef28a6935af9d87a3b9d9d72e2 change added a bug. The original assertion was not the same to `ASSERT(!m_data.empty());`, and which is wrong.

hyjorc1 replied:
Then, maybe this is better since we want to avoid nullptr arithmetic? 
```
ASSERT(m_data.empty() || (position >= m_data.data() && position <= m_data.data() + m_data.size()));
auto newSize = m_data.empty() ? 0 : m_data.data() + m_data.size() - position;
m_data = { position, newSize };
```

Constellation replied:
I don't think this is great since it is adding more costly branches. I think right now I'll keep the current code since this running code is not changed, and this patch's goal is not fixing StringParsingBuffer's potential UB issue introduced by https://github.com/WebKit/WebKit/commit/d2a029f757b316ef28a6935af9d87a3b9d9d72e2

Constellation replied:
I think the right fix is not having nullptr span. When it is set, and it is empty, probably it should define something non null span, like, `m_data { bitwise_cast<T*>(this), 0 }` or something.

MenloDorian reviewed (approved):
r=me

Constellation committed:
Adopt SIMD in attribute and text scanning
https://bugs.webkit.org/show_bug.cgi?id=273977
rdar://127843610

Reviewed by Mark Lam.

This patch integrates SIMD into attribute value and string scanning.
We also attach UNLIKELY / LIKELY to bailout paths.

* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::parseCompleteInput):
(WebCore::HTMLFastPathParser::scanText):
(WebCore::HTMLFastPathParser::scanTagName):
(WebCore::HTMLFastPathParser::scanAttributeValue):
(WebCore::HTMLFastPathParser::parseChildren):
(WebCore::HTMLFastPathParser::parseAttributes):
(WebCore::HTMLFastPathParser::parseContainerElement):

Canonical link: https://commits.webkit.org/278647@main
